### DISCUSSION TITLE
Linker error for debug renderer

### DISCUSSION THREAD

**Discussion by mickeyordog** [jrouwe/JoltPhysics#1410] [Q&A]
Hello, my project compiles fine when running the sample Hello World file, but I'm currently in the process of implementing a debug renderer and I'm getting the following error:
```
[build] Undefined symbols for architecture arm64:
[build]   "typeinfo for JPH::DebugRenderer", referenced from:
[build]       typeinfo for PhysicsDebugRenderer in libphysics.a[3](physicsDebugRenderer.cpp.o)
[build] ld: symbol(s) not found for architecture arm64
```
I'm using macOS Sonoma 14.4.1, in a VS Code project. Jolt is integrated via vcpkg/CMake with vcpkg install joltphysics[core,debugrenderer,profiler], so the debug renderer should be included there. I ran `nm -gC path/to/libJolt.a | grep DebugRenderer` and there were a lot of finds, maybe there's somehow no definition for it though. I'm not positive where to go from here. As mentioned, the default parts of the library are building and linking just fine.

Secondary question, what should the Batch in `Batch CreateTriangleBatch()` contain? I wasn't able to find any documentation about it, from what I see it's just a RefTargetVirtual? Thank you.

**Comment by jrouwe** [comment]
> Undefined symbols for architecture arm64

Jolt by default compiles without C++ RTTI, your project is compiling with RTTI. Either turn it off for your own project or turn it on for Jolt using the CPP_RTTI_ENABLED cmake option.

> Secondary question, what should the Batch in Batch CreateTriangleBatch() contain? I wasn't able to find any documentation about it, from what I see it's just a RefTargetVirtual? Thank you.

It's a handle to renderable data. It can be whatever you need it to be in order to be able to draw a batch of triangles. See example here:

https://github.com/jrouwe/JoltPhysics/blob/9b3e9e604145eee5845edf54fb1fc05585fa61c0/TestFramework/Renderer/DebugRendererImp.h#L61-L71

https://github.com/jrouwe/JoltPhysics/blob/9b3e9e604145eee5845edf54fb1fc05585fa61c0/TestFramework/Renderer/DebugRendererImp.cpp#L98-L107

**Comment by mickeyordog** [comment]
Got it thank you. Is there a way to do that with vcpkg or would I need to compile from source myself?

**Comment by jrouwe** [comment]
I'm not the maintainer of the vcpkg, but looking at the vcpkg repository:

https://github.com/microsoft/vcpkg/blob/ff8d715b6dfcc192f4930e8e696afe6a2e0e85c0/ports/joltphysics/portfile.cmake#L13-L17

it is not exposed. 

Exposing it should be trivial though. Just add something along the lines `rtti_enabled CPP_RTTI_ENABLED` to that code block and send a PR to microsoft.

**Reply by mickeyordog** [reply]
Thank you. I don't really know enough about CMake to feel confident submitted that myself, so I opened a request [here](https://github.com/microsoft/vcpkg/issues/42847). If you would be at all willing to help with this yourself that would be awesome, but otherwise I'll just wait to see if they are able to help with it.

**Reply by mickeyordog** [reply]
By the way I had one more question. For my debug renderer I call initialize in the constructor after the Jolt hello world is set up, and call NextFrame(), but nothing is being displayed. I checked with the debugger and only the initialize and CreateBatch functions are being called, none of the draw functions are being called. Do you have any idea why that might be?

**Reply by jrouwe** [reply]
You need to call `PhysicsSystem::DrawBodies` or activate any of the other debug drawing functions for Jolt to draw anything:

https://github.com/jrouwe/JoltPhysics/blob/248253f88a5945c2032ffb397101afb9f88aeca5/Samples/SamplesApp.cpp#L527-L568
